alerting/src/main/kotlin/org/opensearch/alerting/DocumentReturningMonitorRunner.kt

```diff
@@ -49,37 +50,81 @@ object DocumentReturningMonitorRunner : MonitorRunner {
         val docsToQueries = mutableMapOf<String, MutableList<String>>()
-        val docExecutionContext = DocumentExecutionContext(queries, lastRunContext, updatedLastRunContext)
-        queries.forEach { query ->
-            val matchingDocIds = runForEachQuery(monitorCtx, docExecutionContext, query, index)
-            queryToDocIds[query] = matchingDocIds
-            matchingDocIds.forEach {
-                docsToQueries.putIfAbsent(it, mutableListOf())
-                docsToQueries[it]?.add(query.id)
+
+        try {
+            val getAliasesRequest = GetAliasesRequest(index)
+            val getAliasesResponse = monitorCtx.client!!.admin().indices().getAliases(getAliasesRequest).actionGet()
+            val aliasIndices = getAliasesResponse.aliases.keys().map { it.value }
+
+            // TODO: Add log message if index is/is not alias?
+            val isAlias = aliasIndices.isNotEmpty()
+
+            // If the input index is an alias, creating a list of all indices associated with that alias;
+            // else creating a list containing the single index input
+            val indices = if (isAlias) getAliasesResponse.aliases.keys().map { it.value } else listOf(index)
+
+            indices.forEach { indexName ->
+                // Prepare lastRunContext for each index
+                val indexLastRunContext = lastRunContext.getOrPut(indexName) { createRunContext(monitorCtx, indexName) }
+
+                // Prepare updatedLastRunContext for each index
+                val indexUpdatedRunContext = updateLastRunContext(
+                    indexLastRunContext.toMutableMap(),
+                    monitorCtx,
+                    indexName
+                ) as MutableMap<String, Any>
+                updatedLastRunContext[indexName] = indexUpdatedRunContext
+
+                // Prepare DocumentExecutionContext for each index
+                val docExecutionContext = DocumentExecutionContext(queries, indexLastRunContext, indexUpdatedRunContext)
+
+                val queryResults = getDocLevelQueryResults(
+                    monitorCtx,
+                    docExecutionContext,
+                    indexName,
+                    queries
+                )
+
+                // Only need to compile the query results together when executing the monitor for more than 1 index
+                if (indices.size > 1) {
+                    val newQueryToDocIds = mutableMapOf<DocLevelQuery, Set<String>>()
+                    queryResults["queryDocIds"]?.forEach {
+                        val docLevelQuery = it.key
+                        val matchingDocIds = it.value
+                        val existingQueryToDocIds = queryToDocIds.getOrDefault(docLevelQuery, setOf()).toMutableSet()
+                        existingQueryToDocIds.addAll(matchingDocIds)
+                        newQueryToDocIds[it.key as DocLevelQuery] = existingQueryToDocIds
+                    }
+                    queryToDocIds.putAll(newQueryToDocIds)
```

Comment:
You can replace this logic with this to simplify it:
```
queryToDocIds.apply {
queryResults["queryDocIds"]?.forEach  { (k,v) -> merge(k,v) { oldSet, newSet -> oldSet + newSet } } 
}
```
Once you do this, you dont really need to check if `indices.size > 1` and you can make this logic more general.

Some [reference](https://www.techiedelight.com/merge-two-maps-in-kotlin/).